java/ql/lib/semmle/code/java/dataflow/TypeFlow.qll

```diff
@@ -8,757 +8,359 @@
+   * Holds if data can flow from `n1` to `n2` in one step, and `n1` is
+   * functionally determined by `n2`.
    */
-  private predicate edgeRank2(int r2, int r1, Node n) {
-    r1 = rank[r2](int r | edgeRank1(r, _, n) | r)
+  predicate step(TypeFlowNode n1, TypeFlowNode n2) {
+    n2.asExpr() = n1.asField().getAnAccess()
+    or
+    n2.asExpr() = n1.asSsa().getAUse()
+    or
+    n2.asExpr().(CastingExpr).getExpr() = n1.asExpr() and
+    not n2.asExpr().getType() instanceof PrimitiveType
+    or
+    n2.asExpr().(AssignExpr).getSource() = n1.asExpr() and
+    not n2.asExpr().getType() instanceof PrimitiveType
+    or
+    n2.asSsa().(BaseSsaUpdate).getDefiningExpr().(VariableAssign).getSource() = n1.asExpr()
+    or
+    n2.asSsa().(BaseSsaImplicitInit).captures(n1.asSsa())
   }
 
-  /** Holds if `r` is a ranking of the incoming edges `(n1,n2)` to `n2`. */
-  predicate edgeRank(int r, TypeFlowNode n1, Node n2) {
-    exists(int r1 |
-      edgeRank1(r1, n1, n2) and
-      edgeRank2(r, r1, n2)
+  /**
+   * Holds if `null` is the only value that flows to `n`.
+   */
+  predicate isNullValue(TypeFlowNode n) {
+    n.asExpr() instanceof NullLiteral
+    or
+    exists(LocalVariableDeclExpr decl |
+      n.asSsa().(BaseSsaUpdate).getDefiningExpr() = decl and
+      not decl.hasImplicitInit() and
+      not exists(decl.getInit())
     )
   }
 
-  int lastRank(Node n) { result = max(int r | edgeRank(r, _, n)) }
-}
-
-private signature module TypePropagation {
-  class Typ;
-
-  predicate candType(TypeFlowNode n, Typ t);
+  predicate isExcludedFromNullAnalysis(TypeFlowNode n) {
+    // Fields that are never assigned a non-null value are probably set by
+    // reflection and are thus not always null.
+    exists(n.asField())
+  }
 
-  bindingset[t]
-  predicate supportsType(TypeFlowNode n, Typ t);
-}
+  predicate exactTypeBase(TypeFlowNode n, RefType t) {
+    exists(J::ClassInstanceExpr e |
+      n.asExpr() = e and
+      e.getType() = t and
+      not e instanceof FunctionalExpr and
+      exists(SrcRefType sub | sub.getASourceSupertype() = t.getSourceDeclaration())
+    )
+  }
 
-/** Implements recursion through `forall` by way of edge ranking. */
-private module ForAll<NodeSig Node, RankedEdge<Node> E, TypePropagation T> {
   /**
-   * Holds if `t` is a bound that holds on one of the incoming edges to `n` and
-   * thus is a candidate bound for `n`.
+   * Holds if `n` occurs in a position where type information might be discarded;
+   * `t1` is the type of `n`, `t1e` is the erasure of `t1`, `t2` is the type of
+   * the implicit or explicit cast, and `t2e` is the erasure of `t2`.
    */
   pragma[nomagic]
-  private predicate candJoinType(Node n, T::Typ t) {
-    exists(TypeFlowNode mid |
-      T::candType(mid, t) and
-      E::edgeRank(_, mid, n)
+  private predicate upcastCand(TypeFlowNode n, RefType t1, RefType t1e, RefType t2, RefType t2e) {
+    exists(TypeFlowNode next | step(n, next) or Make<J::Location, Input>::joinStep(n, next) |
```

Comment:
I don't think we need to tie this instantiation-dependent knot here. We may provide a few more base candidates, but they won't affect the recursive derivation of type bounds.
```suggestion
    exists(TypeFlowNode next | step(n, next) or joinStep(n, next) |
```